### DISCUSSION TITLE
v2.0.1 ModuleNotFoundError: 'pip._internal.operations.build' + 'requests'

### DISCUSSION THREAD

**Discussion by 3DSpotters** [derekhe/msfs2020-google-map#28] [General]
v2.0.1 not working on my side (Win10 64):
(run as ADMIN as well as regular user)

...> .\run.bat

D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement>python\python.exe -m pip install -r requirements.txt
Traceback (most recent call last):
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\python\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\python\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\python\lib\site-packages\pip\__main__.py", line 31, in <module>
    sys.exit(_main())
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\python\lib\site-packages\pip\_internal\cli\main.py", line 68, in main
    command = create_command(cmd_name, isolated=("--isolated" in cmd_args))
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\python\lib\site-packages\pip\_internal\commands\__init__.py", line 94, in create_command
    module = importlib.import_module(module_path)
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\python\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\python\lib\site-packages\pip\_internal\commands\install.py", line 14, in <module>
    from pip._internal.cli.req_command import (
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\python\lib\site-packages\pip\_internal\cli\req_command.py", line 21, in <module>
    from pip._internal.index.package_finder import PackageFinder
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\python\lib\site-packages\pip\_internal\index\package_finder.py", line 32, in <module>
    from pip._internal.req import InstallRequirement
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\python\lib\site-packages\pip\_internal\req\__init__.py", line 8, in <module>
    from .req_install import InstallRequirement
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\python\lib\site-packages\pip\_internal\req\req_install.py", line 26, in <module>
    from pip._internal.operations.build.metadata import generate_metadata
**ModuleNotFoundError: No module named 'pip._internal.operations.build'**

D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement>python\python.exe src\app.py
Traceback (most recent call last):
  File "D:\MSFS2020_ADDONS_R\FSUTILS\MSFS2020_Google_Map_replacement\src\app.py", line 6, in <module>
    import requests
**ModuleNotFoundError: No module named 'requests'**

**Comment by 3DSpotters** [comment]
OK, now it works :)
I changed run.bat to:

@setlocal enableextensions
@cd /d "%~dp0"

python.exe -m pip install -r requirements.txt
python.exe src\app.py

==> the problem was with the Python included in the package, my installed before Python 3.9 was OK.

Anyway, I leave this message here, maybe helpful - not everybody can play with BAT files :)

THANKS for such a smart software!

**Reply by tarukuu** [reply]
Thank you . I changed too.

**Comment by 3DSpotters** [comment]
But in the end: Google does not want to send me pictures:

...
requests.exceptions.ConnectionError: SOCKSHTTPSConnectionPool(host='mt1.google.com', port=443): Max retries exceeded with url: /vt/lyrs=s&x=133&y=86&z=8 (Caused by NewConnectionError('<urllib3.contrib.socks.SOCKSHTTPSConnection object at 0x000001709C0EDB20>: Failed to establish a new connection: [WinError 10060] 

The connection attempt failed because the connected party did not respond correctly after the specified time period, or the created connection failed because the connected host did not respond '

...

**Reply by derekhe** [reply]
make sure you setup proper proxy url 
if not use proxy to access google, keep it empty

**Reply by 3DSpotters** [reply]
I am not using any "proxy". I keep the "proxy field" empy, test is OK.
I tried both direct internet access as well as VPN to another country, same  [WinError 10060] ...

**Comment by derekhe** [comment]
Please test the latest version: https://github.com/derekhe/msfs2020-google-map/releases/tag/2.0.3

**Reply by 3DSpotters** [reply]
I started MSFS2020 in the end...and wait :)

1. BTW: run.bat: there is no "py" in my system, I have "pyhon" instead :)
2. Certificate error (warning) during MSFS2020 start...

**Reply by derekhe** [reply]
the "py"  is for default py 3.9 installation, if you install the python again you should have it
certificate error, please click "yes" to allow it.

**Reply by 3DSpotters** [reply]
OK...MSFS2002 still starting, tiles ARE LOADING (in debug CMD window)...

**Reply by 3DSpotters** [reply]
OUCH...my MSFS2020 did CTD (first time for a long time) :(
Before entering the menu...
OK, I will try to start again...

**Reply by derekhe** [reply]
yes, it takes long time to enter the game, I don't know why.

**Reply by 3DSpotters** [reply]
No, this long time is not a problem: this is because lots of ADDONS in the Community folder, this is "OK" anyway.
CTD is not OK :)

**Reply by derekhe** [reply]
what is CTD?

**Reply by 3DSpotters** [reply]
Oh. CRASH. Crash to the Desktop, without any warning.

**Reply by 3DSpotters** [reply]
Now (the second try to start) my MSFS wait in the middle of loading, no sound heard...

**Comment by 3DSpotters** [comment]
Yep, CTD again :(
After first CTD, when I started second time I got RED WARNIG to start in "safe" or in "normal" way. I have chosen "normal". I think "safe" is without ADDONS or without internet?

**Comment by 3DSpotters** [comment]
I will try now to revert to no "Google replacement" and with "hosts" commented.

**Comment by 3DSpotters** [comment]
I reverted to "default" - now I start my MSFS again.

**Comment by 3DSpotters** [comment]
OK, looks like my MSFS in default config (with "Google replacement" switched OFF) will start - now the bar goes right after 50%.
OK, now I am back in the main menu.

To be sure I may do a test with ALL ADDONS OFF in Community folder and only "Google replacement" ON...

**Reply by derekhe** [reply]
ok, may be something wrong. In my PC it sometimes happen. Will try to find out.
If I wait for sometime it will goes in to the menu and can start the game.

**Reply by 3DSpotters** [reply]
So... I do last test for now - "Google replacement" ON and all ADDONS OFF (my Community folder is empty now) ...

**Reply by 3DSpotters** [reply]
My MSFS started with "Google replacement" ON and ALL ADDONS OFF... now I try to fly...

**Comment by 3DSpotters** [comment]
127.0.0.1 - - [15/Sep/2021 17:38:38] "GET /tiles/akh1202033022123003.jpeg?n=z&g=10652 HTTP/1.0" 500 -
Downloading from: https://mt1.google.com/vt/lyrs=s&x=34344&y=22235&z=16 None
Downloading from: https://mt1.google.com/vt/lyrs=s&x=34343&y=22235&z=16 None
Downloading from: https://mt1.google.com/vt/lyrs=s&x=34343&y=22234&z=16 None
...

**Reply by derekhe** [reply]
it seems working

**Reply by 3DSpotters** [reply]
YES. On plain "default" MSFS now I see textures... to be sure I will delete ROLLING CACH and try to spawn somewhery completly else...

**Reply by 3DSpotters** [reply]
Not sure what is going on. I constantly got information about "too low bandwidth" and MSFS goes into "OFFLINE "mode automatically!

**Reply by 3DSpotters** [reply]
I can turn "connection" ON in the in-flight menu but... I still have default MSFS textures, not textures from Google!

**Comment by 3DSpotters** [comment]
"    return session.request(method=method, url=url, **kwargs)
  File "C:\Users\rysio\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\requests\sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "C:\Users\rysio\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\requests\sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "C:\Users\rysio\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\requests\adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='mt1.google.com', port=443): Max retries exceeded with url: /vt/lyrs=s&x=8293&y=5918&z=14 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000002860F7A36A0>: Failed to establish a new connection: [WinError 10060] Próba połączenia nie powiodła się, ponieważ połączona strona nie odpowiedziała poprawnie po ustalonym okresie czasu lub utworzone połączenie nie powiodło się, ponieważ połączony host nie odpowiedział'))
127.0.0.1 - - [15/Sep/2021 17:51:14] "GET /tiles/akh12022201122321.jpeg?n=z&g=9591 HTTP/1.0" 500 -
Downloading from: https://mt1.google.com/vt/lyrs=s&x=8292&y=5914&z=14 None
127.0.0.1 - - [15/Sep/2021 17:51:16] "GET /tiles/akh12022201122302.jpeg?n=z&g=9591 HTTP/1.0" 200 -
Downloading from: https://mt1.google.com/vt/lyrs=s&x=2070&y=1475&z=12 None
127.0.0.1 - - [15/Sep/2021 17:51:18] "GET /tiles/akh120222010132.jpeg?n=z&g=1 HTTP/1.0" 200 -
127.0.0.1 - - [15/Sep/2021 17:51:18] "GET /tiles/akh120220211.jpeg?n=z&g=1 HTTP/1.0" 200 -
Downloading from: https://mt1.google.com/vt/lyrs=s&x=4146&y=2955&z=13 None
Downloading from: https://mt1.google.com/vt/lyrs=s&x=4143&y=2955&z=13 None
127.0.0.1 - - [15/Sep/2021 17:51:18] "GET /tiles/akh1202220112032.jpeg?n=z&g=9591 H"

**Comment by 3DSpotters** [comment]
Still: "Failed to establish a new connection: [WinError 10060]" as you see... and I have  DATA=>ONLINE FUNCTIONALITY ON...

**Comment by 3DSpotters** [comment]
![textures_default](https://user-images.githubusercontent.com/47834083/133467195-6eb1fd88-bb0a-4b4b-90e0-582d769840ea.jpg)

**Reply by 3DSpotters** [reply]
After a while...is it...GOOGLE???
![textures_Google_WOOOW](https://user-images.githubusercontent.com/47834083/133467572-0e9bf8c8-0501-4e83-8b23-c552040c4da4.jpg)

**Reply by 3DSpotters** [reply]
YES! IT IS Google...looks like the mechanism (or Google itself) needs some time (a few minutes?) to "stabilize" on a given area...then the map tiles are sent from Google!
I never before was in the area and DEBUG window shows the tiles are coming from Google...now without any errors...

Tiles are not coming as fast as from BING itself but they come in the end.
I also have to debug my ADDONS - which one causes CTD in pair with "Google replacement" ...

**Comment by 3DSpotters** [comment]
So the conclusion is - I CAN USE Google but:
- a mechanism is not no way as reliable as default BING,
- you have to be patient and fly both HIGH and SLOW :)
- some ADDONS cause CTD in pair with "Google replacement" (they are not causing CTD in default config)
- We can not give any ticket to Google about THIS overall problem with MSFS :D :D  :D

**Comment by 3DSpotters** [comment]
Next test: I turned ON some of ADONS.
In the beginning of the start MSFS gives me a message, that "We can not log you to Xbox Live service.. Press any key to start" (I am on a PC).
When I press a key, the screen is all dark but the music is ON...
I closed the MSFS window and turn OFF and ON "Google replacement" and started MSFS again.
Again the message:
"We can not log you to Xbox Live service. Try in a while (0x80004005)".

So now I switched OFF ALL ADDON SCENERY (and I left all addon aircraft) - looks like MSFS will start, let's hope (no "Xbox Live" service message) ...


HAHAHA - that "Xbox Live" thing was... because current Microsoft Servers outage!
So MSFS started now and I can start my flight. It takes much longer to load tiles from Google but it works in the end.
Looks like that some pre-SU5 sceneries are causing CTD sometimes (they do not without "Google replacement" thing though...).

**Reply by derekhe** [reply]
From your description I can see that your google connection seems not stable and slow.

**Reply by 3DSpotters** [reply]
I wonder what can I do with that...
OK - next step with my testing: I will try VPN to different countries, I hope my VPN has enough bandwidth...

**Reply by 3DSpotters** [reply]
YES!
So I tried in the morning today AND directed my VPN to... New York, USA.
Starting up process was long but when I started my flight...every  texture is Google's and crisp!!!
Looks like both evening and my native location (PL) are not the best for loading Google tiles :)
PS. CMD Debug Window shows now NO (timeout?) ERRORS, every tile loads as it should!

Thank you very much for your consultancy, @derekhe !

![textures_Google_FAST_NOW](https://user-images.githubusercontent.com/47834083/133562506-28c5fe4b-ec95-4ab5-97de-48bdf61467cf.jpg)
